### ISSUE TITLE
Support new failOnDynamicVersionResolution()/failOnChangingModuleResolution() options for the ResolutionStrategy API

### ISSUE THREAD

**Issue by steffenyount** [gradle/gradle#8901]
To support deterministic dependency resolution and build/test reproducibility, we need a way to prevent dependency versions from being selected/resolved dynamically and changing between builds at runtime.

### Expected Behavior
The ResolutionStrategy API could include 2 new settings like:
```
failOnDynamicVersionResolution()
failOnChangingModuleResolution()
```
which would cause a Gradle failure when either dynamic version resolution or changing module resolution was to be used for selecting the resolved dependency artifact.

e.g: with failOnDynamicVersionResolution() set, I would expect that declaring a dependency on `com.fasterxml.jackson.module:jackson-module-guice:2.9.8` which has a transitive dependency on `com.google.inject:guice:[3.0,5.0)` will cause a gradle failure unless I have also declared a dependency on a specific version of that module say `com.google.inject:guice:4.2.0`

### Current Behavior
There is no clean/easy way to do this within the existing Gradle DSL, though it is possible to approximate the desired behavior by throwing an exception when a resolvedDependencyResult's selectionReason's description starts with "didn't match versions", but that doesn't seem to work for all dynamic versions and it's a really hacky solution!

### Context
We're managing a large build system spanning multiple teams and repos, with one of our design goals being: reproducible builds/tests for any given git commit. A consequence of this is that we need our dependency resolution results to be deterministic for any given git commit.

We've been using lockfiles to support deterministic dependency resolution, but that isn't an ideal solution since the update procedure is burdensome on our devs, and it is still possible for any team to introduce dynamic dependencies that can require constant updates to keep the affected lockfiles up-to-date.

To help prevent our devs from shooting themselves in the feet like this, we want to implement a policy that just prohibits any reliance on dynamic dependency resolution in our builds, and fails fast when such a dependency is accidentally added.

**Comment by ljacomet** [comment]
As you stated, if the objective is reproducible build, dependency locking is the answer Gradle has.

I am not sure I understand the burdensome aspect you describe. Moving away from dynamic versions + locking back to single version declarations will move the burden to updating these versions away from updating the lock. I can understand the later feels more familiar initially though.

**Comment by steffenyount** [comment]
Since it's not obvious, let me try to describe the burdensome nature of our dev experience in maintaining the lockfile mechanism.

Our multi-project build currently comprises 431 projects, from 20+ separate teams, requiring 6043 lockfiles be checked-in.

With this quantity and scale of lockfiles it is not realistic to expect our devs to do any manual or targeted editing of these lockfiles. Instead, the common workflow is for devs to simply regenerate a complete replacement set of lockfiles starting from the root project directory, and then to include those lockfile changes in their commits.

This situation is problematic for us in a number of ways:

1. A 1 line change to the dependencies in the wrong project can literally require updates to tens of hundreds of lockfiles because of the transitive dependencies between projects.

2. We're using Bitbucket for our source control, and our code review process is significantly impacted when these overwhelming numbers of lockfile diffs are displayed. We have been hearing tons of complaints about this from our devs, and we have no real work arounds for them. There's no way to automatically hide or minimize just the lockfile diffs, meanwhile our devs are complaining that they're accidentally missing important changes in the PRs they are reviewing, because they have to manually pick the important changes out from so many hundreds of un-informative lockfile diffs.

3. The large surface area presented by the high numbers of lockfile changes in these commits also makes these PRs much more likely to be blocked for merge conflicts. Our devs back this up by grumbling that they are often required to do more than one pass through the whole "merge the latest changes from master", "regenerate lockfiles", "push new changes", "find reviewers for signoff", and "wait for the CI build/test cycle to complete successfully" workflow (for what they feel are trivial dependency changes), because their first attempts are being blocked by merge conflicts before their CI build/tests and reviewers have even had time to finish.  

4. Being that regenerating the lockfiles is a manual step, we've found that this step is also often being missed by devs when they go to push a dependency update to their downstream projects, and as a consequence they waste a bunch of time shepherding an extra PR/merge that has no effect, or worse the devs move-on to doing other work with the false impression that their change is in effect and that their change has passed all the CI tests, while in fact the next unsuspecting dev to regenerate the lockfiles is now on the hook as the first dev to actually test their change.

5. To ensure that this manual step has not been missed and to avoid placing this burden on an unsuspecting downstream dev, we've implemented a pre-commit build pipeline task that ensures the lockfiles are always up-to-date for the declared dependencies of a change when the PRs are submitted. At this point we might as well be automating the process of regenerating the lockfiles to avoid the human errors, but then we're back to non-reproducible builds wherever a project declares a dynamic dependency for resolution.

6. This makes our build pipelines fragile. Since, there's currently no supported Gradle mechanism to enforce a policy which prohibits the use of dynamic dependency resolution, any dev can introduce dynamic dependencies that will then require constant updates to keep the lockfiles current. And since a dev could do this, a dev did do this: We had a dev decide their team's integration tests would depend on the latest build artifacts coming out of an upstream CI build pipeline. They used a dynamic version dependency `4.5.*` for that, and as a consequence the related lockfiles were out-of-date after every build of that upstream CI pipeline. This then caused widespread CI build/test failures for unrelated teams, since the lockfile up-to-date checking pipeline task would detect that the affected lockfiles in all PRs (including PRs with no changes to dependency declarations) were now out-of-date after every run of that upstream CI pipeline.


I hope I have managed to illustrate how the lockfile mechanism is not an adequate solution for us to achieve reproducible builds. The lockfile mechanism does not make it easy to implement reproducible builds, and it does not guide our devs into doing the right thing of declaring reproducibly resolvable dependencies.

What we really want is a policy enforcement mechanism that will prevent devs from introducing any runtime dynamism into the dependency version resolution process. And, if we had had that mechanism to start with we wouldn't have needed to introduce the lockfile mechanism or the burdens that came with it.

**Comment by ljacomet** [comment]
As locking does not seem practical / desired in your setup, here is how you can make sure no dynamic versions enter the graph:

```
plugins {
    `java-library`
}

repositories {
    mavenCentral {
        // The lister referenced here will be called _every time_ a dynamic version is requested
        // if and only if the result is not cached by Gradle. Cache expires after 24h by default
        setComponentVersionsLister(DenyDynamicVersion::class.java)
    }
}

dependencies {

    // Adding the following declaration skips the dynamic resolution
//    implementation("com.google.inject:guice:4.2.2")

    // Adding the following will clean the metadata to remove the dynamic version
//    components {
//        withModule("com.fasterxml.jackson.module:jackson-module-guice", ForceGuiceVersionWhenDynamic::class.java)
//    }

    // A transitive dependency on com.google.inject:guice:[3.0,5.0) will trigger the version lister, failing resolution.
    implementation("com.fasterxml.jackson.module:jackson-module-guice:2.9.8")

    testImplementation("junit:junit:4.12")
}

class DenyDynamicVersion : ComponentMetadataVersionLister {
    override fun execute(details: ComponentMetadataListerDetails) {
        throw IllegalStateException("Using a dynamic version for ${details.moduleIdentifier}")
    }

}

class ForceGuiceVersionWhenDynamic : ComponentMetadataRule {
    override fun execute(t: ComponentMetadataContext) {
        t.details.allVariants {
            withDependencies {
                filter { it.group == "com.google.inject" && it.name == "guice" && it.versionConstraint.displayName.contains("[")}.forEach {
                    it.version { require("4.2.2")}
                    it.because("No dynamic versions are allowed")
                }
            }
        }
    }
}
```

I recommend setting up something like that in your build as there is no plan to add the requested methods.

**Comment by ljacomet** [comment]
For changing version, again a `ComponentMetadataRule` is what you need:

```
dependencies {
    // Adding the following will prevent _changing_ modules in  in direct and transitives dependencies
    components {
        all(DenyChangingModules::class.java)
    }
}

class DenyChangingModules : ComponentMetadataRule {
    override fun execute(t: ComponentMetadataContext) {
        if (t.details.isChanging) {
            throw IllegalStateException("Detected changing module ${t.details.id}")
        }
    }
}
```

Note that `ComponentMetadataRule` are executed in order in which they are added. If another rule in the build changes a component to `changing` _after_ this rule has been applied, then it will not be detected.

**Comment by ljacomet** [comment]
Closing as adding the feature is not planned and solutions have been provided.

**Comment by ljacomet** [comment]
This will be integrated in upcoming Gradle 6.1, see PR #10936 for details

**Comment by melix** [comment]
The PR has been merged. @steffenyount would be great if you could try [this nightly](https://services.gradle.org/distributions-snapshots/gradle-6.1-20191004133052+0000-bin.zip) and let us know if it works for you.

**Comment by steffenyount** [comment]
@melix - Hi!

I wanted to let you know that I'm really excited that this change has made it into the mainline codebase, and I'm looking forward to rolling it out on our large multi-project build!

Unfortunately, that build still needs to be fixed to avoid some deprecated functionality with the move to Gradle 6+ versions. So simply plugging the nightly build into that project for an acceptance test isn't as straightforward as I would like.

Alternatively, I guess I could build a local toy/test project that exercises the new code, but I'm a bit strapped for time right now (probably for the next few weeks), and I assume you've already done something similar in your test codes?

Again, thanks and this is awesome!

**Comment by melix** [comment]
>  I assume you've already done something similar in your test code?

Yes we do, but nothing's better than a real world project to test :)
